ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -176,7 +194,38 @@ Uploader::set_runtime_id(std::string_view id)
 }
 
 bool
+Uploader::thread_upload_impl(const Profile* profile)
+{
+    // As per convention, we wrap the thread so that all return paths
+    // mark the termination of the thread.
+    // There's a race condition here, because the thread will still be
+    // alive for several cycles after the return of this, but it shouldn't
+    // matter since that context is non-consuming and in a pre-join state.
+    bool ret = upload_impl(profile);
+    thread_working.store(false);
+    return ret;
+}
+
+void
 Uploader::upload(const Profile* profile)
+{
+    // If the upload thread is busy, give it some time.  We give it up to 3 seconds, which is arbitrary
+    wait_for_thread(3);
```

Comment:
waiting here means we are at a new cycle and still have not finished uploading?
Would we stop profiling for 3 seconds here ?

Reply:
Exports are scheduled separately from sampling, so this is actually quite complicated.  We'll block the caller, but sampling will proceed.  This means that samples will continue to accrue in the current buffer while the uploader is blocked.